# NHL Game Prediction Modeling
by Gary Schwaeber

## Overview

With sport betting becoming increasingly popular and mainstream I believe that data science can be used to make superior decisions over gut intuitions. In this notebook I will attempt to train logistic regression, ada boost, and gradient boosting models in an attempt to make the best possible game prediction model. I will train my models and tune model hyperparemetres using game results from seasons '2017-2018', '2018-2019', '2019-2020'. Then I will predict on held out games from the current 2021 season and evaluate my model. There are currently a handful of public models whose log loss on the current season's games is being [tracked](https://hockey-statistics.com/2021/05/03/game-projections-january-13th-2021/) on which I can compare the quality of my model to. The score I will look to optimize is log loss, however, I will also review accuracy scores due to their interpretability.

Log-loss is indicative of how close the prediction probability is to the corresponding actual/true value (0 or 1 in case of binary classification). The more the predicted probability diverges from the actual value, the higher is the log-loss value. [Source](https://towardsdatascience.com/intuition-behind-log-loss-score-4e0c9979680a)


In [158]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import numpy as np
import statsmodels.api as sm
import hockey_scraper
import pickle
import time
import random
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.preprocessing import normalize, FunctionTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss, accuracy_score
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.metrics import roc_auc_score, roc_curve, auc

from sklearn.metrics import confusion_matrix, plot_confusion_matrix,\
    precision_score, recall_score, accuracy_score, f1_score, log_loss,\
    roc_curve, roc_auc_score, classification_report
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier, AdaBoostClassifier, AdaBoostRegressor, GradientBoostingClassifier
from collections import Counter
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.compose import ColumnTransformer
from sklearn.datasets import fetch_openml
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.compose import make_column_selector as selector
from sklearn.pipeline import make_pipeline, Pipeline
from sklearn.feature_selection import RFECV
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv('data/all_games_multirolling_SVA.csv')

In [3]:
conditions = [((df['date'] >= '2017-10-04') & (df['date'] <= '2018-04-08')),
              ((df['date'] >= '2018-10-03') & (df['date'] <= '2019-04-06')),
              ((df['date'] >= '2019-10-02') & (df['date'] <= '2020-03-12')),
              ((df['date'] >= '2021-01-13') & (df['date'] <= '2021-04-29'))
             ]
, 
choices = ['2017-2018',
           '2018-2019',
           '2019-2020',
           '2020-2021']
           
    

df['Season'] = np.select(conditions, choices)

In [245]:
# define feature columns for different rolling intervals
r3 = ['home_B2B', 'away_B2B', 'home_Goalie_FenwickSV%',
 'home_Goalie_GSAx/60',
 'home_Goalie_HDCSV%',
 'away_Goalie_FenwickSV%',
 'away_Goalie_GSAx/60',
 'away_Goalie_HDCSV%', 'home_last_3_FF%_5v5',
 'home_last_3_GF%_5v5',
 'home_last_3_xGF%_5v5',
 'home_last_3_SH%',
 'home_last3_pp_TOI_per_game',
 'home_last3_xGF_per_min_pp',
 'home_last3_pk_TOI_per_game',
 'home_last3_xGA_per_min_pk', 'away_last_3_FF%_5v5',
 'away_last_3_GF%_5v5',
 'away_last_3_xGF%_5v5',
 'away_last_3_SH%',
 'away_last3_pp_TOI_per_game',
 'away_last3_xGF_per_min_pp',
 'away_last3_pk_TOI_per_game',
 'away_last3_xGA_per_min_pk']
r5 =['home_Goalie_FenwickSV%',
 'home_Goalie_GSAx/60',
 'home_Goalie_HDCSV%',
 'away_Goalie_FenwickSV%',
 'away_Goalie_GSAx/60',
 'away_Goalie_HDCSV%', 'home_B2B', 'away_B2B', 'home_last_5_FF%_5v5',
 'home_last_5_GF%_5v5',
 'home_last_5_xGF%_5v5',
 'home_last_5_SH%',
 'home_last5_pp_TOI_per_game',
 'home_last5_xGF_per_min_pp',
 'home_last5_pk_TOI_per_game',
 'home_last5_xGA_per_min_pk', 'away_last_5_FF%_5v5',
 'away_last_5_GF%_5v5',
 'away_last_5_xGF%_5v5',
 'away_last_5_SH%',
 'away_last5_pp_TOI_per_game',
 'away_last5_xGF_per_min_pp',
 'away_last5_pk_TOI_per_game',
 'away_last5_xGA_per_min_pk']
r10 =['home_Goalie_FenwickSV%',
 'home_Goalie_GSAx/60',
 'home_Goalie_HDCSV%',
 'away_Goalie_FenwickSV%',
 'away_Goalie_GSAx/60',
 'away_Goalie_HDCSV%', 'home_B2B', 'away_B2B', 'home_last_10_FF%_5v5',
 'home_last_10_GF%_5v5',
 'home_last_10_xGF%_5v5',
 'home_last_10_SH%',
 'home_last10_pp_TOI_per_game',
 'home_last10_xGF_per_min_pp',
 'home_last10_pk_TOI_per_game',
 'home_last10_xGA_per_min_pk', 'away_last_10_FF%_5v5',
 'away_last_10_GF%_5v5',
 'away_last_10_xGF%_5v5',
 'away_last_10_SH%',
 'away_last10_pp_TOI_per_game',
 'away_last10_xGF_per_min_pp',
 'away_last10_pk_TOI_per_game',
 'away_last10_xGA_per_min_pk']
r20 = ['home_Goalie_FenwickSV%',
 'home_Goalie_GSAx/60',
 'home_Goalie_HDCSV%',
 'away_Goalie_FenwickSV%',
 'away_Goalie_GSAx/60',
 'away_Goalie_HDCSV%', 'home_B2B', 'away_B2B',  'home_last_20_FF%_5v5',
 'home_last_20_GF%_5v5',
 'home_last_20_xGF%_5v5',
 'home_last_20_SH%',
 'home_last20_pp_TOI_per_game',
 'home_last20_xGF_per_min_pp',
 'home_last20_pk_TOI_per_game',
 'home_last20_xGA_per_min_pk', 'away_last_20_FF%_5v5',
 'away_last_20_GF%_5v5',
 'away_last_20_xGF%_5v5',
 'away_last_20_SH%',
 'away_last20_pp_TOI_per_game',
 'away_last20_xGF_per_min_pp',
 'away_last20_pk_TOI_per_game',
 'away_last20_xGA_per_min_pk']
r30 = ['home_Goalie_FenwickSV%',
 'home_Goalie_GSAx/60',
 'home_Goalie_HDCSV%',
 'away_Goalie_FenwickSV%',
 'away_Goalie_GSAx/60',
 'away_Goalie_HDCSV%', 'home_B2B', 'away_B2B',  'home_last_30_FF%_5v5',
 'home_last_30_GF%_5v5',
 'home_last_30_xGF%_5v5',
 'home_last_30_SH%',
 'home_last30_pp_TOI_per_game',
 'home_last30_xGF_per_min_pp',
 'home_last30_pk_TOI_per_game',
 'home_last30_xGA_per_min_pk', 'away_last_30_FF%_5v5',
 'away_last_30_GF%_5v5',
 'away_last_30_xGF%_5v5',
 'away_last_30_SH%',
 'away_last30_pp_TOI_per_game',
 'away_last30_xGF_per_min_pp',
 'away_last30_pk_TOI_per_game',
 'away_last30_xGA_per_min_pk']
r40 = ['home_Goalie_FenwickSV%',
 'home_Goalie_GSAx/60',
 'home_Goalie_HDCSV%',
 'away_Goalie_FenwickSV%',
 'away_Goalie_GSAx/60',
 'away_Goalie_HDCSV%', 'home_B2B', 'away_B2B',
'home_last_40_FF%_5v5',
 'home_last_40_GF%_5v5',
 'home_last_40_xGF%_5v5',
 'home_last_40_SH%',
 'home_last40_pp_TOI_per_game',
 'home_last40_xGF_per_min_pp',
 'home_last40_pk_TOI_per_game',
 'home_last40_xGA_per_min_pk',
'away_last_40_FF%_5v5',
 'away_last_40_GF%_5v5',
 'away_last_40_xGF%_5v5',
 'away_last_40_SH%',
 'away_last40_pp_TOI_per_game',
 'away_last40_xGF_per_min_pp',
 'away_last40_pk_TOI_per_game',
 'away_last40_xGA_per_min_pk']
all_r = list(set(r3+r5+r10+r20+r30+r40))

r3_30 =list(set(r3+r30))
r5_30 = list(set(r5+r30))
r10_30 = list(set(r10+r30))
r_3_5_30 = list(set(r3+r5+r30))
r_5_20 = list(set(r5+r20))
r_5_40 = list(set(r5+r40))

## Baseline Model

The baseline model will predict that every home team wins their game and that the probability of that is the ratio of games the home team has won.

In [374]:
df['Home_Team_Won'].value_counts(normalize=True)

1    0.541458
0    0.458542
Name: Home_Team_Won, dtype: float64

In [371]:
baseline_preds = np.ones(df.shape[0])
accuracy_score(df['Home_Team_Won'],baseline_preds)

0.5414581066376496

In [381]:
baseline_probs = np.repeat(df['Home_Team_Won'].value_counts(normalize=True)[1], df.shape[0])

log_loss(df['Home_Team_Won'], baseline_probs)

0.689705681560888

The models will need to beat an accuracy score of 54.15% and a log loss of .6897, otherwise they are no better than just predicting the home team will win. 

## Rolling 5 and 40 game features

For my first set of models I will attempt using 5 and 40 game rolling features. These seemed like a good set based on the feature selection notebook. 40 games is currently the longest rolling runway I have currently for the team statistics. The 40 games stats intuitively provide the most smoothing of team data over the course of the season, while the 5 game stats may provide some insight on any streakiness or may cover recent developments that would affect short term team performances such as player injuries, trades coaching changes etc.

In [21]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,r_5_40]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].loc[:,r_5_40]
y_test = df[df['Season'] == '2020-2021']['Home_Team_Won']

In [23]:
X_train.columns

Index(['home_last_5_FF%_5v5', 'home_last5_xGF_per_min_pp',
       'home_last40_pp_TOI_per_game', 'away_last40_pk_TOI_per_game',
       'home_last5_pk_TOI_per_game', 'away_B2B', 'away_last40_xGF_per_min_pp',
       'home_Goalie_GSAx/60', 'home_last_5_SH%', 'away_last5_pk_TOI_per_game',
       'away_last_5_GF%_5v5', 'away_Goalie_GSAx/60', 'home_last_40_GF%_5v5',
       'away_last_5_xGF%_5v5', 'home_B2B', 'away_last5_xGF_per_min_pp',
       'home_last40_pk_TOI_per_game', 'away_last_40_SH%',
       'away_last_40_GF%_5v5', 'home_last_40_xGF%_5v5',
       'home_Goalie_FenwickSV%', 'home_last5_xGA_per_min_pk',
       'home_last_5_GF%_5v5', 'away_Goalie_HDCSV%', 'home_last_40_SH%',
       'away_last_40_xGF%_5v5', 'away_last40_pp_TOI_per_game',
       'home_last40_xGA_per_min_pk', 'home_last5_pp_TOI_per_game',
       'away_last_5_FF%_5v5', 'away_last5_xGA_per_min_pk',
       'home_last_40_FF%_5v5', 'away_last5_pp_TOI_per_game',
       'home_last40_xGF_per_min_pp', 'home_last_5_xGF%_5v5', 'away_

In [31]:
numeric_features = ['home_last_5_FF%_5v5', 'home_last5_xGF_per_min_pp',
       'home_last40_pp_TOI_per_game', 'away_last40_pk_TOI_per_game',
       'home_last5_pk_TOI_per_game', 'away_last40_xGF_per_min_pp',
       'home_Goalie_GSAx/60', 'home_last_5_SH%', 'away_last5_pk_TOI_per_game',
       'away_last_5_GF%_5v5', 'away_Goalie_GSAx/60', 'home_last_40_GF%_5v5',
       'away_last_5_xGF%_5v5', 'away_last5_xGF_per_min_pp',
       'home_last40_pk_TOI_per_game', 'away_last_40_SH%',
       'away_last_40_GF%_5v5', 'home_last_40_xGF%_5v5',
       'home_Goalie_FenwickSV%', 'home_last5_xGA_per_min_pk',
       'home_last_5_GF%_5v5', 'away_Goalie_HDCSV%', 'home_last_40_SH%',
       'away_last_40_xGF%_5v5', 'away_last40_pp_TOI_per_game',
       'home_last40_xGA_per_min_pk', 'home_last5_pp_TOI_per_game',
       'away_last_5_FF%_5v5', 'away_last5_xGA_per_min_pk',
       'home_last_40_FF%_5v5', 'away_last5_pp_TOI_per_game',
       'home_last40_xGF_per_min_pp', 'home_last_5_xGF%_5v5', 'away_last_5_SH%',
       'away_last40_xGA_per_min_pk', 'away_Goalie_FenwickSV%',
       'away_last_40_FF%_5v5', 'home_Goalie_HDCSV%']

In [26]:
X_train.loc[:, numerical_columns]

home_last_5_FF%_5v5  home_last5_xGF_per_min_pp  \
0               52.399869                   0.079714   
1               42.564205                   0.143856   
2               60.511924                   0.113316   
3               60.511924                   0.113316   
4               54.316401                   0.118615   
...                   ...                        ...   
3790            52.626239                   0.059924   
3791            43.058811                   0.096857   
3792            48.469552                   0.102022   
3793            55.838089                   0.086931   
3794            46.028554                   0.102724   

      home_last40_pp_TOI_per_game  away_last40_pk_TOI_per_game  \
0                        5.328333                     4.540000   
1                        4.705417                     4.928750   
2                        4.682500                     5.185417   
3                        4.682500                     5.185417   
4                        4.778333                     5.305000   
...                           ...                          ...   
3790                     5.620833                     5.130833   
3791                     4.900833                     4.127083   
3792                     4.770833                     4.755833   
3793                     5.192083                     5.041667   
3794                     3.902917                     4.673750   

      home_last5_pk_TOI_per_game  away_last40_xGF_per_min_pp  \
0                       3.693333                    0.122400   
1                       3.546667                    0.102018   
2                       4.540000                    0.120843   
3                       4.540000                    0.120843   
4                       4.763333                    0.143998   
...                          ...                         ...   
3790                    4.210000                    0.112455   
3791                    5.073333                    0.136510   
3792                    5.786667                    0.122391   
3793                    5.200000                    0.121545   
3794                    5.043333                    0.119628   

      home_Goalie_GSAx/60  home_last_5_SH%  away_last5_pk_TOI_per_game  \
0               -0.334940         9.426112                    3.070000   
1                0.205712        12.093988                    4.966667   
2                0.312441         8.478124                    5.853333   
3                0.312441         8.478124                    5.853333   
4               -0.232180         9.804628                    5.963333   
...                   ...              ...                         ...   
3790            -0.004131         9.856137                    4.503333   
3791            -0.646591        12.419735                    3.396667   
3792            -0.435356        11.374701                    3.583333   
3793            -0.029116         9.257587                    3.623333   
3794             0.211833         8.850640                    5.180000   

      away_last_5_GF%_5v5  away_Goalie_GSAx/60  home_last_40_GF%_5v5  \
0               45.937500             0.027934             50.127801   
1               49.927641            -0.138771             56.868932   
2               45.427286             0.041876             56.575634   
3               45.427286             0.041876             56.575634   
4               56.272661             0.009622             53.260259   
...                   ...                  ...                   ...   
3790            50.523013            -0.160907             50.702434   
3791            70.421512             0.072030             47.486397   
3792            53.981623             0.135063             43.586998   
3793            49.006951            -0.054492             56.092965   
3794            47.310888            -0.483024             46.737740   

      away_last_5

In [113]:
scoring = ['neg_log_loss', 'accuracy']

In [155]:
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())])

categorical_features = ['home_B2B', 'away_B2B']


preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', 'passthrough', categorical_features)])

### Logistic Regression

In [118]:
log_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('logisticregression', LogisticRegression(max_iter=10000))])

log_params = {'logisticregression__solver' : ['liblinear', 'lbfgs', 'newton-cg'],
                'logisticregression__penalty': ['l1', 'l2'],
                'logisticregression__C': [0.1, 10, 20, 100],
                'logisticregression__class_weight': [None] }

log_cv = GridSearchCV(log_pipeline, param_grid=log_params, cv=5, scoring=scoring, refit = 'neg_log_loss',  verbose=1)

In [119]:
log_cv.fit(X_train, y_train)

Fitting 5 folds for each of 24 candidates, totalling 120 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         ['home_last_5_FF%_5v5',
                                                                          'home_last5_xGF_per_min_pp',
                                                                          'home_last40_pp_TOI_per_game',
                                                                          'away_last40_pk_TOI_per_game',
                                                                          'home_last5_pk_TOI_per_game',
                                                                          'away_last40_xGF_per_min_pp',
                

In [120]:
log_cv.best_score_

-0.6777347180223572

In [194]:
log_results = pd.DataFrame(log_cv.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
log_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.027825      0.003296         0.009434        0.000387   
3        0.023710      0.001415         0.008098        0.000967   
5        0.024544      0.001020         0.007175        0.000070   
4        0.017814      0.000625         0.007292        0.000235   
6        0.042029      0.003558         0.007798        0.000512   
12       0.037968      0.004370         0.007319        0.000199   
9        0.024154      0.001340         0.007429        0.000419   
11       0.024628      0.000918         0.007113        0.000054   
10       0.018765      0.000731         0.007361        0.000541   
18       0.039950      0.002819         0.007582        0.000092   

   param_logisticregression__C param_logisticregression__class_weight  \
0                          0.1                                   None   
3                          0.1                                   None   
5                          0.1                                   None   
4                          0.1                                   None   
6                           10                                   None   
12                          20                                   None   
9                           10                                   None   
11                          10                                   None   
10                          10                                   None   
18                         100                                   None   

   param_logisticregression__penalty param_logisticregression__solver  \
0                                 l1                        liblinear   
3                                 l2                        liblinear   
5                                 l2                        newton-cg   
4                                 l2                            lbfgs   
6                                 l1                        liblinear   
12                                l1                        liblinear   
9                                 l2                        liblinear   
11                                l2                        newton-cg   
10                                l2                            lbfgs   
18                                l1                        liblinear   

                                               params  \
0   {'logisticregression__C': 0.1, 'logisticregres...   
3   {'logisticregression__C': 0.1, 'logisticregres...   
5   {'logisticregression__C': 0.1, 'logisticregres...   
4   {'logisticregression__C': 0.1, 'logisticregres...   
6   {'logisticregression__C': 10, 'logisticregress...   
12  {'logisticregression__C': 20, 'logisticregress...   
9   {'logisticregression__C': 10, 'logisticregress...   
11  {'logisticregression__C': 10, 'logisticregress...   
10  {'logisticregression__C': 10, 'logisticregress...   
18  {'logisticregression__C': 100, 'logisticregres...   

    split0_test_neg_log_loss  split1_test_neg_log_loss  \
0                  -0.677067                 -0.675773   
3                  -0.675531                 -0.677123   
5                  -0.675577                 -0.677124   
4                  -0.675577                 -0.677125   
6                  -0.675282                 -0.677541   
12                 -0.675278                 -0.677565   
9                  -0.675276                 -0.677583   
11                 -0.675276                 -0.677584   
10                 -0.675279                 -0.677585   
18                 -0.675276                 -0.677585   

    split2_test_neg_log_loss  split3_test_neg_log_loss  \
0                  -0.682986                 -0.677729   
3                  -0.683068                 -0.676456   
5                  -0.683118                 -0.676415   
4                  -0.683118                 -0.676414   
6                  -0.683798                 -0.676713   
12                 -0.683819    

### Ada Boost

In [151]:
ada_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('ada', AdaBoostClassifier())])

ada_params = {'ada__n_estimators': [25, 50],
         'ada__learning_rate': [.1, 1, 10, 20],
         'ada__base_estimator': [svm.SVC(probability=True , kernel='linear'), LogisticRegression()],}

ada_cv = GridSearchCV(ada_pipeline, param_grid=ada_params, cv=5, scoring=scoring, refit='neg_log_loss', verbose=1)

In [152]:
ada_cv.fit(X_train, y_train)

Fitting 5 folds for each of 16 candidates, totalling 80 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         ['home_last_5_FF%_5v5',
                                                                          'home_last5_xGF_per_min_pp',
                                                                          'home_last40_pp_TOI_per_game',
                                                                          'away_last40_pk_TOI_per_game',
                                                                          'home_last5_pk_TOI_per_game',
                                                                          'away_last40_xGF_per_min_pp',
                

In [153]:
ada_cv.best_score_

-0.6812883430589551

In [156]:
ada_results = pd.DataFrame(ada_cv.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
ada_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
4      48.373599      0.345728         2.937553        0.031958   
0      51.813747      0.627942         2.980223        0.038913   
8       0.144892      0.005547         0.017487        0.001032   
6      44.035846      2.534538         2.475855        0.240989   
5      88.420303      5.276782         5.004536        0.535126   
1      95.772775      2.769245         5.614874        0.184574   
7      81.078607      7.450631         4.446727        0.699147   
9       0.292306      0.043241         0.027016        0.002916   
2      39.229604      0.798806         2.387321        0.059241   
3      75.819306      1.990608         4.576631        0.098662   

                param_ada__base_estimator param_ada__learning_rate  \
4  SVC(kernel='linear', probability=True)                       10   
0  SVC(kernel='linear', probability=True)                      0.1   
8                    LogisticRegression()                      0.1   
6  SVC(kernel='linear', probability=True)                       20   
5  SVC(kernel='linear', probability=True)                       10   
1  SVC(kernel='linear', probability=True)                      0.1   
7  SVC(kernel='linear', probability=True)                       20   
9                    LogisticRegression()                      0.1   
2  SVC(kernel='linear', probability=True)                        1   
3  SVC(kernel='linear', probability=True)                        1   

  param_ada__n_estimators                                             params  \
4                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
0                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
8                      25  {'ada__base_estimator': LogisticRegression(), ...   
6                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
5                      50  {'ada__base_estimator': SVC(kernel='linear', p...   
1                      50  {'ada__base_estimator': SVC(kernel='linear', p...   
7                      50  {'ada__base_estimator': SVC(kernel='linear', p...   
9                      50  {'ada__base_estimator': LogisticRegression(), ...   
2                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
3                      50  {'ada__base_estimator': SVC(kernel='linear', p...   

   split0_test_neg_log_loss  split1_test_neg_log_loss  \
4                 -0.682460                 -0.678853   
0                 -0.682541                 -0.680892   
8                 -0.684177                 -0.683080   
6                 -0.685870                 -0.680804   
5                 -0.686099                 -0.680916   
1                 -0.685556                 -0.684639   
7                 -0.687720                 -0.681945   
9                 -0.686973                 -0.686153   
2                 -0.688021                 -0.688112   
3                 -0.688886                 -0.688712   

   split2_test_neg_log_loss  split3_test_neg_log_loss  \
4                 -0.681992                 -0.682383   
0                 -0.682944                 -0.680844   
8                 -0.685022                 -0.682982   
6                 -0.684960                 -0.684752   
5                 -0.685079                 -0.684745   
1                 -0.685505                 -0.684347   
7                 -0.686726                 -0.686418   
9                 -0.687641                 -0.686227   
2                 -0.688802                 -0.688062   
3                 -0.689001                 -0.688580   

   split4_test_neg_log_loss  mean_test_neg_log_loss  std_test_neg_log_loss  \
4                 -0.680754               -0.681288               0.001363   
0                 -0.680954               -0.681635               0.000914   
8                 -0.683228               -0.683698               0.000787   
6                 -0.684487               -0.

### Gradient Boosting

In [143]:
gb_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('gb', GradientBoostingClassifier())])

gb_params = {'gb__n_estimators': [200, 300, 400],
         'gb__learning_rate': [.001,.01, .1],
         'gb__max_depth' : [3,5]}

gb_cv = GridSearchCV(gb_pipeline, param_grid=gb_params, cv=5, scoring=scoring, refit='neg_log_loss', verbose=1)

In [144]:
gb_cv.fit(X_train, y_train)

Fitting 5 folds for each of 18 candidates, totalling 90 fits
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=200; total time=   4.9s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=200; total time=   5.0s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=200; total time=   5.1s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=200; total time=   5.1s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=200; total time=   5.1s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=300; total time=   7.6s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=300; total time=   7.5s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=300; total time=   7.4s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=300; total time=   7.4s
[CV] END gb__learning_rate=0.001, gb__max_depth=3, gb__n_estimators=300; total time=   7.5s
[CV] END gb__learni

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         ['home_last_5_FF%_5v5',
                                                                          'home_last5_xGF_per_min_pp',
                                                                          'home_last40_pp_TOI_per_game',
                                                                          'away_last40_pk_TOI_per_game',
                                                                          'home_last5_pk_TOI_per_game',
                                                                          'away_last40_xGF_per_min_pp',
                

In [145]:
gb_cv.best_score_

-0.6813076251374858

In [146]:
gb_results = pd.DataFrame(gb_cv.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
gb_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
6       5.271923      0.141695         0.019683        0.009091   
7       7.873014      0.096896         0.016901        0.003443   
8      10.301410      0.380511         0.015873        0.000352   
5      16.534748      0.199185         0.027762        0.001951   
2      10.285022      0.335853         0.019065        0.001787   
9       7.998002      0.027494         0.016288        0.000805   
1       7.484383      0.079754         0.015311        0.000887   
4      12.391780      0.195259         0.024282        0.005705   
0       5.027624      0.073685         0.013116        0.001382   
3       8.214938      0.091112         0.019273        0.000652   

  param_gb__learning_rate param_gb__max_depth param_gb__n_estimators  \
6                    0.01                   3                    200   
7                    0.01                   3                    300   
8                    0.01                   3                    400   
5                   0.001                   5                    400   
2                   0.001                   3                    400   
9                    0.01                   5                    200   
1                   0.001                   3                    300   
4                   0.001                   5                    300   
0                   0.001                   3                    200   
3                   0.001                   5                    200   

                                              params  \
6  {'gb__learning_rate': 0.01, 'gb__max_depth': 3...   
7  {'gb__learning_rate': 0.01, 'gb__max_depth': 3...   
8  {'gb__learning_rate': 0.01, 'gb__max_depth': 3...   
5  {'gb__learning_rate': 0.001, 'gb__max_depth': ...   
2  {'gb__learning_rate': 0.001, 'gb__max_depth': ...   
9  {'gb__learning_rate': 0.01, 'gb__max_depth': 5...   
1  {'gb__learning_rate': 0.001, 'gb__max_depth': ...   
4  {'gb__learning_rate': 0.001, 'gb__max_depth': ...   
0  {'gb__learning_rate': 0.001, 'gb__max_depth': ...   
3  {'gb__learning_rate': 0.001, 'gb__max_depth': ...   

   split0_test_neg_log_loss  split1_test_neg_log_loss  \
6                 -0.682190                 -0.680045   
7                 -0.683010                 -0.680717   
8                 -0.684514                 -0.682343   
5                 -0.684924                 -0.684652   
2                 -0.685048                 -0.685482   
9                 -0.683699                 -0.684479   
1                 -0.685524                 -0.686056   
4                 -0.685632                 -0.685264   
0                 -0.686338                 -0.686972   
3                 -0.686655                 -0.686523   

   split2_test_neg_log_loss  split3_test_neg_log_loss  \
6                 -0.683195                 -0.679583   
7                 -0.683615                 -0.679676   
8                 -0.684595                 -0.679947   
5                 -0.683665                 -0.685081   
2                 -0.683626                 -0.685462   
9                 -0.685518                 -0.683762   
1                 -0.684620                 -0.686407   
4                 -0.685363                 -0.685987   
0                 -0.685883                 -0.687305   
3                 -0.685808                 -0.687157   

   split4_test_neg_log_loss  mean_test_neg_log_loss  std_test_neg_log_loss  \
6                 -0.681526               -0.681308               0.001338   
7                 -0.680791               -0.681561               0.001495   
8                 -0.681175               -0.682515               0.001830   
5                 -0.686546               -0.684974               0.000928   
2                 -0.685492               -0.685022               0.000718   
9                 -0.688825               -0.685257               0.001901   
1                 -0.685681               -0.685658  

It does not seem that gradient boosting is producing good results for this dataset

### Feature Importance Evaluation

Reviewing the logistic regression, coefficients, I can see which feature the algorithm deemed most impactful. I am
very surprised that away_last_40_xGF%_5v5 was cut by the l1 regularization, that seemed like it would be one of the more important features.

In [191]:
log_coef = pd.DataFrame(list(zip(X_train.columns, log_cv.best_estimator_[1].coef_[0])), columns = ['Feature', 'Coef'] )
log_coef['Coef_abs'] = abs(log_coef['Coef'])
log_coef.sort_values('Coef_abs', ascending = False)

Feature      Coef  Coef_abs
18         away_last_40_GF%_5v5  0.190469  0.190469
36   away_last40_xGA_per_min_pk -0.159196  0.159196
38         away_last_40_FF%_5v5 -0.152225  0.152225
29          away_last_5_FF%_5v5  0.151841  0.151841
10          away_last_5_GF%_5v5 -0.067696  0.067696
39           home_Goalie_HDCSV%  0.060833  0.060833
15    away_last5_xGF_per_min_pp -0.060642  0.060642
14                     home_B2B  0.060171  0.060171
17             away_last_40_SH%  0.057695  0.057695
3   away_last40_pk_TOI_per_game -0.054198  0.054198
37       away_Goalie_FenwickSV% -0.047271  0.047271
5                      away_B2B -0.046307  0.046307
21    home_last5_xGA_per_min_pk -0.045549  0.045549
20       home_Goalie_FenwickSV% -0.042995  0.042995
11          away_Goalie_GSAx/60 -0.038014  0.038014
16  home_last40_pk_TOI_per_game -0.036960  0.036960
19        home_last_40_xGF%_5v5 -0.035819  0.035819
22          home_last_5_GF%_5v5  0.033234  0.033234
31         home_last_40_FF%_5v5  0.032698  0.032698
33   home_last40_xGF_per_min_pp -0.032378  0.032378
26  away_last40_pp_TOI_per_game -0.025321  0.025321
2   home_last40_pp_TOI_per_game  0.024676  0.024676
24             home_last_40_SH% -0.020544  0.020544
23           away_Goalie_HDCSV% -0.017224  0.017224
0           home_last_5_FF%_5v5  0.012531  0.012531
7           home_Goalie_GSAx/60 -0.009327  0.009327
4    home_last5_pk_TOI_per_game -0.009273  0.009273
1     home_last5_xGF_per_min_pp -0.007824  0.007824
30    away_last5_xGA_per_min_pk -0.004043  0.004043
34         home_last_5_xGF%_5v5  0.003836  0.003836
12         home_last_40_GF%_5v5  0.002082  0.002082
13         away_last_5_xGF%_5v5 -0.000514  0.000514
32   away_last5_pp_TOI_per_game  0.000000  0.000000
28   home_last5_pp_TOI_per_game  0.000000  0.000000
25        away_last_40_xGF%_5v5  0.000000  0.000000
35              away_last_5_SH%  0.000000  0.000000
6    away_last40_xGF_per_min_pp  0.000000  0.000000
8               home_last_5_SH%  0.000000  0.000000
9    away_last5_pk_TOI_per_game  0.000000  0.000000
27   home_last40_xGA_per_min_pk  0.000000  0.000000

## 40 Game Rolling

I will run some models using only the rolling 40 game team stats

In [351]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,r40]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].loc[:,r40]
y_test = df[df['Season'] == '2020-2021']['Home_Team_Won']

In [352]:
numeric_features = [
       'home_last40_pp_TOI_per_game', 'away_last40_pk_TOI_per_game',
 'away_last40_xGF_per_min_pp',
       'home_Goalie_GSAx/60',  'away_Goalie_GSAx/60', 'home_last_40_GF%_5v5',
       'home_last40_pk_TOI_per_game', 'away_last_40_SH%',
       'away_last_40_GF%_5v5', 'home_last_40_xGF%_5v5',
       'home_Goalie_FenwickSV%',
       'away_Goalie_HDCSV%', 'home_last_40_SH%',
       'away_last_40_xGF%_5v5', 'away_last40_pp_TOI_per_game',
       'home_last40_xGA_per_min_pk', 
       'home_last_40_FF%_5v5', 
       'home_last40_xGF_per_min_pp',
       'away_last40_xGA_per_min_pk', 'away_Goalie_FenwickSV%',
       'away_last_40_FF%_5v5', 'home_Goalie_HDCSV%']

### Logistic Regression

In [209]:
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())])

categorical_features = ['home_B2B', 'away_B2B']


preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', 'passthrough', categorical_features)])

log_40_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('logisticregression', LogisticRegression(max_iter=10000))])

In [210]:
log_params = {'logisticregression__solver' : ['liblinear', 'lbfgs', 'newton-cg'],
                'logisticregression__penalty': ['l1', 'l2'],
                'logisticregression__C': [.01, 0.1, 1, 10],
                'logisticregression__class_weight': [None] }

log_cv_40 = GridSearchCV(log_40_pipeline, param_grid=log_params, cv=5, scoring=scoring, refit = 'neg_log_loss',  verbose=1)

In [211]:
log_cv_40.fit(X_train, y_train)

Fitting 5 folds for each of 24 candidates, totalling 120 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         ['home_last40_pp_TOI_per_game',
                                                                          'away_last40_pk_TOI_per_game',
                                                                          'away_last40_xGF_per_min_pp',
                                                                          'home_Goalie_GSAx/60',
                                                                          'away_Goalie_GSAx/60',
                                                                          'home_last_40_GF%_5v5',
                            

In [212]:
log_40_results = pd.DataFrame(log_cv_40.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
log_40_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
4        0.014693      0.001130         0.007716        0.000719   
5        0.017283      0.000292         0.006951        0.000118   
3        0.014061      0.000746         0.008523        0.001029   
9        0.014426      0.000387         0.006950        0.000158   
11       0.020913      0.000814         0.007221        0.000453   
10       0.015567      0.000349         0.007100        0.000190   
12       0.022668      0.001238         0.007185        0.000171   
15       0.015133      0.000416         0.007112        0.000265   
17       0.021284      0.000212         0.007193        0.000235   
16       0.016655      0.000304         0.006913        0.000190   

   param_logisticregression__C param_logisticregression__class_weight  \
4                         0.01                                   None   
5                         0.01                                   None   
3                         0.01                                   None   
9                          0.1                                   None   
11                         0.1                                   None   
10                         0.1                                   None   
12                           1                                   None   
15                           1                                   None   
17                           1                                   None   
16                           1                                   None   

   param_logisticregression__penalty param_logisticregression__solver  \
4                                 l2                            lbfgs   
5                                 l2                        newton-cg   
3                                 l2                        liblinear   
9                                 l2                        liblinear   
11                                l2                        newton-cg   
10                                l2                            lbfgs   
12                                l1                        liblinear   
15                                l2                        liblinear   
17                                l2                        newton-cg   
16                                l2                            lbfgs   

                                               params  \
4   {'logisticregression__C': 0.01, 'logisticregre...   
5   {'logisticregression__C': 0.01, 'logisticregre...   
3   {'logisticregression__C': 0.01, 'logisticregre...   
9   {'logisticregression__C': 0.1, 'logisticregres...   
11  {'logisticregression__C': 0.1, 'logisticregres...   
10  {'logisticregression__C': 0.1, 'logisticregres...   
12  {'logisticregression__C': 1, 'logisticregressi...   
15  {'logisticregression__C': 1, 'logisticregressi...   
17  {'logisticregression__C': 1, 'logisticregressi...   
16  {'logisticregression__C': 1, 'logisticregressi...   

    split0_test_neg_log_loss  split1_test_neg_log_loss  \
4                  -0.675677                 -0.672521   
5                  -0.675677                 -0.672520   
3                  -0.675375                 -0.672548   
9                  -0.674011                 -0.672891   
11                 -0.674061                 -0.672896   
10                 -0.674062                 -0.672900   
12                 -0.673682                 -0.672968   
15                 -0.673574                 -0.673067   
17                 -0.673579                 -0.673068   
16                 -0.673578                 -0.673068   

    split2_test_neg_log_loss  split3_test_neg_log_loss  \
4                  -0.678186                 -0.674674   
5                  -0.678187                 -0.674673   
3                  -0.678078                 -0.674946   
9                  -0.678995                 -0.674600   
11                 -0.679035                 -0.674563   
10                 -0.679034    

#### Feature Importance Evaluation

In [214]:
log_40_coef = pd.DataFrame(list(zip(X_train.columns, log_cv_40.best_estimator_[1].coef_[0])), columns = ['Feature', 'Coef'] )
log_40_coef['Coef_abs'] = abs(log_40_coef['Coef'])
log_40_coef.sort_values('Coef_abs', ascending = False)

Feature      Coef  Coef_abs
10        home_last_40_xGF%_5v5  0.167769  0.167769
16         away_last_40_FF%_5v5  0.128198  0.128198
22  away_last40_pk_TOI_per_game -0.127853  0.127853
20  away_last40_pp_TOI_per_game -0.120332  0.120332
9          home_last_40_GF%_5v5  0.092343  0.092343
5            away_Goalie_HDCSV% -0.069468  0.069468
7                      away_B2B -0.069465  0.069465
23   away_last40_xGA_per_min_pk  0.068317  0.068317
4           away_Goalie_GSAx/60 -0.058731  0.058731
6                      home_B2B  0.057827  0.057827
1           home_Goalie_GSAx/60 -0.057538  0.057538
2            home_Goalie_HDCSV% -0.053817  0.053817
11             home_last_40_SH% -0.049700  0.049700
13   home_last40_xGF_per_min_pp -0.048419  0.048419
21   away_last40_xGF_per_min_pp -0.047514  0.047514
8          home_last_40_FF%_5v5 -0.044988  0.044988
12  home_last40_pp_TOI_per_game  0.042033  0.042033
17         away_last_40_GF%_5v5  0.037274  0.037274
14  home_last40_pk_TOI_per_game -0.031585  0.031585
0        home_Goalie_FenwickSV%  0.028338  0.028338
3        away_Goalie_FenwickSV%  0.019926  0.019926
15   home_last40_xGA_per_min_pk -0.017327  0.017327
18        away_last_40_xGF%_5v5  0.014227  0.014227
19             away_last_40_SH% -0.008973  0.008973

### Ada Boost

In [354]:
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())])

categorical_features = ['home_B2B', 'away_B2B']


preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', 'passthrough', categorical_features)])

log_40_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('logisticregression', LogisticRegression(max_iter=10000))])

ada_40_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('ada', AdaBoostClassifier())])

ada_params = {'ada__n_estimators': [25],
         'ada__learning_rate': [.01, .1, 1, 10],
         'ada__base_estimator': [svm.SVC(probability=True , kernel='linear'), LogisticRegression()],}

ada_cv_40 = GridSearchCV(ada_40_pipeline, param_grid=ada_params, cv=5, scoring=scoring, refit='neg_log_loss', verbose=1)

In [355]:
ada_cv_40.fit(X_train, y_train)

Fitting 5 folds for each of 8 candidates, totalling 40 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         ['home_last40_pp_TOI_per_game',
                                                                          'away_last40_pk_TOI_per_game',
                                                                          'away_last40_xGF_per_min_pp',
                                                                          'home_Goalie_GSAx/60',
                                                                          'away_Goalie_GSAx/60',
                                                                          'home_last_40_GF%_5v5',
                            

In [357]:
ada_40_results = pd.DataFrame(ada_cv_40.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
ada_40_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0      40.928906      1.102974         2.342911        0.052230   
4       0.123535      0.004822         0.016900        0.001988   
3      42.345423      0.336290         2.573474        0.028111   
1      42.433099      0.796355         2.499568        0.082708   
5       0.132891      0.010832         0.016125        0.000969   
2      35.091805      0.160695         2.052148        0.016464   
6       0.101714      0.006457         0.016900        0.000865   
7       0.271557      0.016521         0.017777        0.002543   

                param_ada__base_estimator param_ada__learning_rate  \
0  SVC(kernel='linear', probability=True)                     0.01   
4                    LogisticRegression()                     0.01   
3  SVC(kernel='linear', probability=True)                       10   
1  SVC(kernel='linear', probability=True)                      0.1   
5                    LogisticRegression()                      0.1   
2  SVC(kernel='linear', probability=True)                        1   
6                    LogisticRegression()                        1   
7                    LogisticRegression()                       10   

  param_ada__n_estimators                                             params  \
0                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
4                      25  {'ada__base_estimator': LogisticRegression(), ...   
3                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
1                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
5                      25  {'ada__base_estimator': LogisticRegression(), ...   
2                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
6                      25  {'ada__base_estimator': LogisticRegression(), ...   
7                      25  {'ada__base_estimator': LogisticRegression(), ...   

   split0_test_neg_log_loss  split1_test_neg_log_loss  \
0                 -0.676974                 -0.673666   
4                 -0.681029                 -0.678332   
3                 -0.683231                 -0.672194   
1                 -0.682249                 -0.680174   
5                 -0.684313                 -0.682629   
2                 -0.688593                 -0.688348   
6                 -0.691506                 -0.691278   
7                 -0.690214                 -0.693413   

   split2_test_neg_log_loss  split3_test_neg_log_loss  \
0                 -0.677154                 -0.676621   
4                 -0.681242                 -0.678748   
3                 -0.682686                 -0.680888   
1                 -0.681820                 -0.680632   
5                 -0.684777                 -0.683140   
2                 -0.688968                 -0.688236   
6                 -0.691629                 -0.691432   
7                 -0.700866                 -0.689553   

   split4_test_neg_log_loss  mean_test_neg_log_loss  std_test_neg_log_loss  \
0                 -0.672943               -0.675472               0.001793   
4                 -0.678069               -0.679484               0.001367   
3                 -0.680843               -0.679968               0.004002   
1                 -0.680885               -0.681152               0.000768   
5                 -0.682848               -0.683541               0.000848   
2                 -0.687525               -0.688334               0.000476   
6                 -0.691436               -0.691456               0.000114   
7                 -0.694382               -0.693686               0.004031   

   rank_test_neg_log_loss  split0_test_accuracy  split1_test_accuracy  \
0                       1              0.560000              0.590667   
4                       2              0.566667              0.577333   
3                       3              0.573333              0.597333   
1                       

## All Rolling Game Features With Recursive Feature Elimination

In [382]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,all_r]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].loc[:,all_r]
y_test = df[df['Season'] == '2020-2021']['Home_Team_Won']

In [383]:
X_train.shape

(3750, 104)

### Recursive Feature Elimination

In [275]:
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())])

categorical_features = ['home_B2B', 'away_B2B']


preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', 'passthrough', categorical_features)])

rfecv = RFECV(estimator= LogisticRegression(max_iter =10000, penalty = 'l2', solver='liblinear', C=.1), step=1, scoring='accuracy')
rfecv_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                      ('rfecv', rfecv)])

In [276]:
rfecv_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('scaler',
                                                                   StandardScaler())]),
                                                  ['home_Goalie_FenwickSV%',
                                                   'home_Goalie_GSAx/60',
                                                   'home_Goalie_HDCSV%',
                                                   'away_Goalie_FenwickSV%',
                                                   'away_Goalie_GSAx/60',
                                                   'away_Goalie_HDCSV%',
                                                   'home_last_3_FF%_5v5',
                                                   'home_last_3_GF%_5v5',
                                                   'home_last_3_xGF%_5v5',
                                                   'home_last_3_SH%',
    

In [278]:
rfecv_pipeline[1].n_features_

23

In [279]:
rfecv_pipeline[1].ranking_

array([ 1,  2, 11, 75,  1, 27,  1, 48,  1, 24, 47, 81,  9, 65, 72, 16, 70,
       14, 37, 34, 42, 77,  1, 67,  1, 23, 46, 62, 10, 12, 45, 15, 36, 13,
       71, 33, 43, 63, 64, 79, 41, 82, 76, 26, 40, 73, 44, 54, 35, 74, 38,
       32, 61, 21, 28, 78, 30, 80, 57, 25,  5,  1,  1, 53,  1,  1, 49, 31,
        8, 69,  1, 66, 29, 52, 51, 55, 59,  1, 19, 18,  6,  1, 50,  1,  7,
       68,  1,  1,  1,  1, 60, 56,  4,  3, 20, 17,  1,  1, 39,  1, 58, 22,
        1,  1])

In [281]:
rfecv_results = pd.DataFrame(list(zip(X_train.columns, rfecv_pipeline[1].ranking_)), columns = ['Feature', 'Ranking']).sort_values('Ranking')
rfecv_results.head(rfecv_pipeline[1].n_features_)

Feature  Ranking
0           home_last_10_FF%_5v5        1
102         home_last_30_FF%_5v5        1
61     away_last3_xGF_per_min_pp        1
62           away_Goalie_GSAx/60        1
64          away_last_5_xGF%_5v5        1
65              away_last_40_SH%        1
70           home_last_3_GF%_5v5        1
77          away_last_40_FF%_5v5        1
24        away_Goalie_FenwickSV%        1
81   away_last40_pk_TOI_per_game        1
86        home_Goalie_FenwickSV%        1
87              away_last_20_SH%        1
88     home_last5_xGA_per_min_pk        1
89              away_last_10_SH%        1
96         away_last_10_xGF%_5v5        1
97    away_last5_pp_TOI_per_game        1
99               away_last_5_SH%        1
83   away_last30_pk_TOI_per_game        1
22          home_last_5_xGF%_5v5        1
103           home_Goalie_HDCSV%        1
6            away_last_3_FF%_5v5        1
4                home_last_3_SH%        1
8    away_last10_pk_TOI_per_game        1

In [282]:
rfecv_columns = list(rfecv_results.iloc[:rfecv_pipeline[1].n_features_,0])
rfecv_columns 

['home_last_10_FF%_5v5',
 'home_last_30_FF%_5v5',
 'away_last3_xGF_per_min_pp',
 'away_Goalie_GSAx/60',
 'away_last_5_xGF%_5v5',
 'away_last_40_SH%',
 'home_last_3_GF%_5v5',
 'away_last_40_FF%_5v5',
 'away_Goalie_FenwickSV%',
 'away_last40_pk_TOI_per_game',
 'home_Goalie_FenwickSV%',
 'away_last_20_SH%',
 'home_last5_xGA_per_min_pk',
 'away_last_10_SH%',
 'away_last_10_xGF%_5v5',
 'away_last5_pp_TOI_per_game',
 'away_last_5_SH%',
 'away_last30_pk_TOI_per_game',
 'home_last_5_xGF%_5v5',
 'home_Goalie_HDCSV%',
 'away_last_3_FF%_5v5',
 'home_last_3_SH%',
 'away_last10_pk_TOI_per_game']

### Logistic Regression

In [283]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,rfecv_columns]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].loc[:,rfecv_columns]
y_test = df[df['Season'] == '2020-2021']['Home_Team_Won']

In [284]:
log_rfecv_pipeline = Pipeline(steps=[('ss', StandardScaler()),
                      ('logisticregression', LogisticRegression(max_iter=10000))])

log_params = {'logisticregression__solver' : ['liblinear', 'lbfgs', 'newton-cg'],
                'logisticregression__penalty': ['l1', 'l2'],
                'logisticregression__C': [.01, 0.1, 10, 20, 100],
                'logisticregression__class_weight': [None]}

log_cv_all = GridSearchCV(log_rfecv_pipeline, param_grid=log_params, cv=5, scoring=scoring, refit = 'neg_log_loss',  verbose=1)

In [285]:
log_cv_all.fit(X_train[rfecv_columns], y_train)

Fitting 5 folds for each of 30 candidates, totalling 150 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('logisticregression',
                                        LogisticRegression(max_iter=10000))]),
             param_grid={'logisticregression__C': [0.01, 0.1, 10, 20, 100],
                         'logisticregression__class_weight': [None],
                         'logisticregression__penalty': ['l1', 'l2'],
                         'logisticregression__solver': ['liblinear', 'lbfgs',
                                                        'newton-cg']},
             refit='neg_log_loss', scoring=['neg_log_loss', 'accuracy'],
             verbose=1)

In [286]:
log_all_results = pd.DataFrame(log_cv_all.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
log_all_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
6        0.011894      0.001292         0.003923        0.000006   
5        0.015744      0.000855         0.004152        0.000450   
4        0.013048      0.001829         0.005018        0.000662   
3        0.010138      0.000692         0.004599        0.000393   
9        0.010136      0.000326         0.003898        0.000107   
11       0.016889      0.000644         0.003922        0.000050   
10       0.012270      0.000551         0.003819        0.000032   
12       0.014359      0.000719         0.004021        0.000203   
18       0.014439      0.000322         0.004291        0.000485   
24       0.017296      0.002053         0.004062        0.000077   

   param_logisticregression__C param_logisticregression__class_weight  \
6                          0.1                                   None   
5                         0.01                                   None   
4                         0.01                                   None   
3                         0.01                                   None   
9                          0.1                                   None   
11                         0.1                                   None   
10                         0.1                                   None   
12                          10                                   None   
18                          20                                   None   
24                         100                                   None   

   param_logisticregression__penalty param_logisticregression__solver  \
6                                 l1                        liblinear   
5                                 l2                        newton-cg   
4                                 l2                            lbfgs   
3                                 l2                        liblinear   
9                                 l2                        liblinear   
11                                l2                        newton-cg   
10                                l2                            lbfgs   
12                                l1                        liblinear   
18                                l1                        liblinear   
24                                l1                        liblinear   

                                               params  \
6   {'logisticregression__C': 0.1, 'logisticregres...   
5   {'logisticregression__C': 0.01, 'logisticregre...   
4   {'logisticregression__C': 0.01, 'logisticregre...   
3   {'logisticregression__C': 0.01, 'logisticregre...   
9   {'logisticregression__C': 0.1, 'logisticregres...   
11  {'logisticregression__C': 0.1, 'logisticregres...   
10  {'logisticregression__C': 0.1, 'logisticregres...   
12  {'logisticregression__C': 10, 'logisticregress...   
18  {'logisticregression__C': 20, 'logisticregress...   
24  {'logisticregression__C': 100, 'logisticregres...   

    split0_test_neg_log_loss  split1_test_neg_log_loss  \
6                  -0.676998                 -0.674722   
5                  -0.677915                 -0.675348   
4                  -0.677915                 -0.675348   
3                  -0.677768                 -0.675230   
9                  -0.678761                 -0.675954   
11                 -0.678788                 -0.675981   
10                 -0.678788                 -0.675980   
12                 -0.678976                 -0.676168   
18                 -0.678992                 -0.676181   
24                 -0.679006                 -0.676194   

    split2_test_neg_log_loss  split3_test_neg_log_loss  \
6                  -0.678140                 -0.677774   
5                  -0.678426                 -0.678199   
4                  -0.678427                 -0.678200   
3                  -0.678595                 -0.678217   
9                  -0.678914                 -0.681263   
11                 -0.678899    

### Ada Boost

In [398]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,rfecv_columns]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].loc[:,rfecv_columns]
y_test = df[df['Season'] == '2020-2021']['Home_Team_Won']

In [399]:
ada_rfecv_pipeline = Pipeline(steps=[('ss', StandardScaler()),
                      ('ada', AdaBoostClassifier())])

ada_params = {'ada__n_estimators': [25],
         'ada__learning_rate': [ .1, 10],
         'ada__base_estimator': [svm.SVC(probability=True , kernel='linear'), LogisticRegression(max_iter =10000, C=.01, penalty = 'l1', solver = 'liblinear')],}

ada_cv_all = GridSearchCV(ada_rfecv_pipeline, param_grid=ada_params, cv=5, scoring=scoring, refit='neg_log_loss', verbose=1)

In [400]:
ada_cv_all.fit(X_train, y_train)

Fitting 5 folds for each of 4 candidates, totalling 20 fits


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('ada', AdaBoostClassifier())]),
             param_grid={'ada__base_estimator': [SVC(kernel='linear',
                                                     probability=True),
                                                 LogisticRegression(C=0.01,
                                                                    max_iter=10000,
                                                                    penalty='l1',
                                                                    solver='liblinear')],
                         'ada__learning_rate': [0.1, 10],
                         'ada__n_estimators': [25]},
             refit='neg_log_loss', scoring=['neg_log_loss', 'accuracy'],
             verbose=1)

In [402]:
ada_all_results = pd.DataFrame(ada_cv_all.cv_results_).sort_values('mean_test_neg_log_loss', ascending=False)
ada_all_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0      45.457442      0.866537         2.623058        0.066013   
1      43.603495      0.634419         2.648859        0.125019   
2       0.077140      0.000653         0.011736        0.000212   
3       0.075963      0.001387         0.012140        0.000086   

                           param_ada__base_estimator param_ada__learning_rate  \
0             SVC(kernel='linear', probability=True)                      0.1   
1             SVC(kernel='linear', probability=True)                       10   
2  LogisticRegression(C=0.01, max_iter=10000, pen...                      0.1   
3  LogisticRegression(C=0.01, max_iter=10000, pen...                       10   

  param_ada__n_estimators                                             params  \
0                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
1                      25  {'ada__base_estimator': SVC(kernel='linear', p...   
2                      25  {'ada__base_estimator': LogisticRegression(C=0...   
3                      25  {'ada__base_estimator': LogisticRegression(C=0...   

   split0_test_neg_log_loss  split1_test_neg_log_loss  \
0                 -0.683233                 -0.681924   
1                 -0.684232                 -0.681059   
2                 -0.693147                 -0.693147   
3                 -0.693147                 -0.693147   

   split2_test_neg_log_loss  split3_test_neg_log_loss  \
0                 -0.683197                 -0.682110   
1                 -0.684624                 -0.682336   
2                 -0.693147                 -0.693147   
3                 -0.693147                 -0.693147   

   split4_test_neg_log_loss  mean_test_neg_log_loss  std_test_neg_log_loss  \
0                 -0.681630               -0.682418               0.000668   
1                 -0.683243               -0.683099               0.001294   
2                 -0.693147               -0.693147               0.000000   
3                 -0.693147               -0.693147               0.000000   

   rank_test_neg_log_loss  split0_test_accuracy  split1_test_accuracy  \
0                       1              0.561333              0.565333   
1                       2              0.569333              0.598667   
2                       3              0.457333              0.457333   
3                       3              0.457333              0.457333   

   split2_test_accuracy  split3_test_accuracy  split4_test_accuracy  \
0              0.541333                 0.568              0.554667   
1              0.553333                 0.568              0.557333   
2              0.457333                 0.456              0.456000   
3              0.457333                 0.456              0.456000   

   mean_test_accuracy  std_test_accuracy  rank_test_accuracy  
0            0.558133           0.009526                   2  
1            0.569333           0.015889                   1  
2            0.456800           0.000653                   3  
3            0.456800           0.000653                   3

## Apply Best Model To Test

I will evaluate the best model iterations on the held out 2021 season data

In [358]:
results_dict = {'cv accuracy': {}, 'cv log loss': {}, 'test accuracy': {}, 'test log_loss':{}}
accuracy_list = []
log_loss_list = []

In [359]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,r_5_40]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].dropna().loc[:,r_5_40]
y_test = df[df['Season'] == '2020-2021'].dropna()['Home_Team_Won']



test_preds_5_40 = log_cv.predict(X_test)

test_probs_5_40 = log_cv.predict_proba(X_test)


accuracy_list.append(accuracy_score(y_test, test_preds_5_40))
log_loss_list.append(log_loss(y_test, test_probs_5_40))


In [360]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,r40]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].dropna().loc[:,r40]
y_test = df[df['Season'] == '2020-2021'].dropna()['Home_Team_Won']



test_preds_40 = log_cv_40.predict(X_test)

test_probs_40 = log_cv_40.predict_proba(X_test)

accuracy_list.append(accuracy_score(y_test, test_preds_40))
log_loss_list.append(log_loss(y_test, test_probs_40))

In [361]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,rfecv_columns]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].dropna().loc[:,rfecv_columns]
y_test = df[df['Season'] == '2020-2021'].dropna()['Home_Team_Won']

test_preds_rfecv = log_cv_all.predict(X_test)

test_probs_rfecv = log_cv_all.predict_proba(X_test)


accuracy_list.append(accuracy_score(y_test, test_preds_rfecv))
log_loss_list.append(log_loss(y_test, test_probs_rfecv))



In [364]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,r_5_40]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].dropna().loc[:,r_5_40]
y_test = df[df['Season'] == '2020-2021'].dropna()['Home_Team_Won']



accuracy_list.append(accuracy_score(y_test, ada_cv.predict(X_test)))
log_loss_list.append(log_loss(y_test,ada_cv.predict_proba(X_test)))

In [365]:
X_train = df[df['Season'] != '2020-2021'].dropna().loc[:,r40]
y_train = df[df['Season'] != '2020-2021'].dropna()['Home_Team_Won']
X_test = df[df['Season'] == '2020-2021'].dropna().loc[:,r40]
y_test = df[df['Season'] == '2020-2021'].dropna()['Home_Team_Won']



accuracy_list.append(accuracy_score(y_test, ada_cv_40.predict(X_test)))
log_loss_list.append(log_loss(y_test, ada_cv_40.predict_proba(X_test)))

In [403]:
results_dict['test accuracy'] = accuracy_list
results_dict['test log_loss'] = log_loss_list
models = ['5 and 40 log', '40 log', 'rfecv log', '5 and 40 ada', '40 ada']
results_dict['cv accuracy'] = [log_results['mean_test_accuracy'][0], log_40_results['mean_test_accuracy'][0], log_all_results['mean_test_accuracy'][0], ada_results['mean_test_accuracy'][0], ada_40_results['mean_test_accuracy'][0]]
results_dict['cv log loss'] = [log_cv.best_score_*-1, log_cv_40.best_score_*-1, log_cv_all.best_score_*-1, ada_cv.best_score_*-1, ada_cv_40.best_score_*-1]

In [404]:
results_df = pd.DataFrame(results_dict, index = models)

## Conclusion

Best model was logistic regression with the rolling 5 and 40 features on the test data. Interestingly, this was the 4th best model on the CV training data set though it did have the best CV accuracy.

In [405]:
results_df.sort_values('test log_loss')

cv accuracy  cv log loss  test accuracy  test log_loss
5 and 40 log     0.581067     0.677735       0.597205       0.657201
40 log           0.580267     0.674081       0.593393       0.657240
40 ada           0.579467     0.675472       0.606099       0.660695
rfecv log        0.571733     0.675783       0.590851       0.665077
5 and 40 ada     0.562667     0.681288       0.560356       0.678121

## Next Steps
To further improve the models I would like to take the following next steps

- Train a neural network model
- Categorize B2B better
- Include team ELO feature
- Try linear weightings in rolling features
- Increase goalie games
- Add prior year goalie GAR feature
- Add Team HDSC % feature
- Add more seasons to training set
- Compare against historical implied odds from a bookmaker
- Adjust ineperienced goalie imputed stats and exclude 2021 season to avoid data leakage on test set